pkg/distributor/query.go

```diff
@@ -53,6 +54,32 @@ func (d *Distributor) Query(ctx context.Context, from, to model.Time, matchers .
 	return matrix, err
 }
 
+func (d *Distributor) QueryExemplars(ctx context.Context, from, to model.Time, matchers ...[]*labels.Matcher) (*ingester_client.QueryResponse, error) {
+	var result *ingester_client.QueryResponse
+	err := instrument.CollectedRequest(ctx, "Distributor.QueryExemplars", d.queryDuration, instrument.ErrorCode, func(ctx context.Context) error {
+		req, err := ingester_client.ToExemplarQueryRequest(from, to, matchers...)
+		if err != nil {
+			return err
+		}
+
+		replicationSet, err := d.GetIngestersForQuery(ctx, false, nil)
```

Comment:
I think it would be good to clarify how all matchers are being sent to all ingesters for the query, even though on ingest they are sharded by series (i.e. it is expected that an ingester will satisfy only a subset of the matchers and the results will be merged).  Maybe some comments, or a different name for the `exemplarQuery` parameter would be enough.

Reply:
> I'm also now thinking of whether we should just split this up more, getting the replica set for each set of matchers and running the exemplar queries with those, and then merging the results.

I don't think it's worth the effort. Let me explain why.

Regardless of shuffle-sharding (which is about tenants sharding, not series), series can be sharded across ingesters into two ways:
1. By metric name only
2. By all series

When sharding by all series you have to query all ingesters, while when sharding by metric name all series for a given metric name are sharded always to the same ingester (+ replicas). That's why we can restrict the ingesters to query from when "shard by metric name only" is used.

However, the current implementation of "query only a restricted set of ingesters when shard by metric name only is enabled" is also buggy and doesn't take in account reshardings (eg. scale up), so it's not even safe to use out of the box with the blocks storage.

Long story short, I think it's fine (and desired) to always query all ingesters for the exemplars. When shuffle-sharding is enabled, it will only query ingesters belonging to the tenant's shard (and that's OK).

Reply:
@pracucci Thanks for the explanation! So in either scenario there's never 3 ingesters that have all of a given tenants data? A set of three would have a subset of their data based on the sharding type?

Reply:
Not sure I understand your question. Could you elaborate more your question, please?